tower-http/src/set_response_header.rs

```diff
@@ -0,0 +1,454 @@
+//!                     // set the `Content-Length` header to that value.
+//!                     Some(HeaderValue::from_str(&size.to_string()).unwrap())
+//!                 } else {
+//!                     // If the response body doesn't have a known size, return `None`
+//!                     // to skip setting the header on this response.
+//!                     None
+//!                 }
+//!             }
+//!         )
+//!     )
+//!     .service(render_html);
+//!
+//! let request = Request::new(Body::empty());
+//!
+//! let response = svc.ready_and().await?.call(request).await?;
+//!
+//! assert_eq!(response.headers()["content-length"], "10");
+//! #
+//! # Ok(())
+//! # }
+//! ```
+
+use futures_util::ready;
+use http::{header::HeaderName, HeaderValue, Response};
+use pin_project::pin_project;
+use std::{
+    fmt,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use std::{future::Future, marker::PhantomData};
+use tower_layer::Layer;
+use tower_service::Service;
+
+/// Layer that applies [`SetResponseHeader`] which adds a response header.
+///
+/// See [`SetResponseHeader`] for more details.
+pub struct SetResponseHeaderLayer<M, T> {
+    header_name: HeaderName,
+    make: M,
+    mode: InsertHeaderMode,
+    _marker: PhantomData<fn() -> T>,
+}
+
+impl<M, T> fmt::Debug for SetResponseHeaderLayer<M, T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("SetResponseHeaderLayer")
+            .field("header_name", &self.header_name)
+            .field("mode", &self.mode)
+            .field("make", &std::any::type_name::<M>())
+            .finish()
+    }
+}
+
+impl<M, T> SetResponseHeaderLayer<M, T>
+where
+    M: MakeHeaderValue<T>,
+{
+    /// Create a new [`SetResponseHeaderLayer`]. If a previous value exists for the same header, it
+    /// is removed and replaced with the new header value.
+    pub fn overriding(header_name: HeaderName, make: M) -> Self {
+        Self::new(header_name, make, InsertHeaderMode::Override)
+    }
+
+    /// Create a new [`SetResponseHeaderLayer`]. The new header is always added, preserving any
+    /// existing values. If previous values exist, the header will have multiple values.
+    pub fn appending(header_name: HeaderName, make: M) -> Self {
+        Self::new(header_name, make, InsertHeaderMode::Append)
+    }
+
+    /// Create a new [`SetResponseHeaderLayer`]. If a previous value exists for the header, the new
+    /// value is not inserted.
+    pub fn if_not_present(header_name: HeaderName, make: M) -> Self {
+        Self::new(header_name, make, InsertHeaderMode::IfNotPresent)
+    }
+
+    fn new(header_name: HeaderName, make: M, mode: InsertHeaderMode) -> Self
+    where
+        M: MakeHeaderValue<T>,
+    {
+        Self {
+            make,
+            header_name,
+            mode,
+            _marker: PhantomData,
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+#[non_exhaustive]
```

Comment:
nit/TIOLI: since this is no longer public we can just remove the non_exhaustive, which is only necessary when the enum is part of the public API stability guarantee.
```suggestion
```